pandas/_libs/tslibs/timestamps.pyx

```diff
@@ -1556,17 +1539,76 @@ class Timestamp(_Timestamp):
         #   check that the second argument is an int.
         #
         # - Nones for the first four (legacy) arguments indicate pydatetime
-        #   keyword arguments. year, month, and day are required. As a
-        #   shortcut, we just check that the first argument was not passed.
-        #
-        # Mixing pydatetime positional and keyword arguments is forbidden!
+        #   keyword arguments. year, month, and day are required. We just
+        #   check that no positional arguments were passed.
 
         cdef:
+            object ts_input=_no_input
             _TSObject ts
-            tzinfo_type tzobj
+            tzinfo_type tzinfo, tzobj
 
-        _date_attributes = [year, month, day, hour, minute, second,
-                            microsecond, nanosecond]
+        args_len = len(args)
+
+        # GH 30543 if pd.Timestamp already passed, return it
+        # check that only ts_input is passed
+        # checking verbosely, because cython doesn't optimize
+        # list comprehensions (as of cython 0.29.x)
+        if args_len == 1 and len(kwargs) == 0 and isinstance(args[0], _Timestamp):
+            return args[0]
+
+        # Building from ts_input
+        if args_len == 1:
+            if kwargs:
+                if ("year" in kwargs or "month" in kwargs or "day" in kwargs or
+                        "hour" in kwargs or "minute" in kwargs or "second" in kwargs or
+                        "microsecond" in kwargs):
+                    raise ValueError("Cannot pass a date attribute keyword argument")
+                if isinstance(args[0], str) and "nanosecond" in kwargs:
+                    raise ValueError(
+                        "Cannot pass a date attribute keyword "
+                        "argument when passing a date string; 'tz' is keyword-only"
+                    )
+
+            ts_input = args[0]
+            tzinfo = kwargs.get("tzinfo")
+        # Building from positional arguments
+        elif 9 > args_len > 2 and isinstance(args[1], int):
+            args = args + (None,) * (8 - args_len)
+            year, month, day, hour, minute, second, microsecond, tzinfo = args
+
+            if kwargs:
+                # Positional or keyword arguments
+                hour = kwargs.get("hour", hour)
```

Comment:
Shouldn't we raise if a user passes the same argument by position and keyword e.g.

```
In [4]: datetime.datetime(2020, 1, 1, day=1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[4], line 1
----> 1 datetime.datetime(2020, 1, 1, day=1)

TypeError: argument for function given by name ('day') and position (3)
```